File: app/src/main/java/com/bigscreen/mangindo/newrelease/NewReleaseActivity.java
package com.bigscreen.mangindo.newrelease;

import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.support.v4.content.ContextCompat;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.RecyclerView;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.KeyEvent;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.view.inputmethod.EditorInfo;
import android.widget.EditText;
import android.widget.FrameLayout;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.bigscreen.mangindo.R;
import com.bigscreen.mangindo.base.BaseActivity;
import com.bigscreen.mangindo.chapter.ChapterListActivity;
import com.bigscreen.mangindo.common.IntentKey;
import com.bigscreen.mangindo.listener.OnListItemClickListener;
import com.bigscreen.mangindo.listener.OnLoadDataListener;
import com.bigscreen.mangindo.network.service.MangaApiService;
import com.bigscreen.mangindo.stored.StoredDataService;

import javax.inject.Inject;

public class NewReleaseActivity extends BaseActivity implements OnLoadDataListener, OnListItemClickListener {

    private SwipeRefreshLayout layoutSwipeRefresh;
    private RecyclerView gridMangaNewRelease;
    private ProgressBar progressLoading;
    private NewReleaseAdapter newReleaseAdapter;

    private FrameLayout layoutSearch;
    private EditText inputSearch;
    private Animation animSlideUp;
    private Animation animSlideDown;

    private boolean isLoading;

    @Inject
    StoredDataService storedDataService;

    @Inject
    MangaApiService mangaApiService;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        getAppDeps().inject(this);
        setContentView(R.layout.activity_new_release);

        if (getSupportActionBar() != null)
            getSupportActionBar().setDisplayShowTitleEnabled(false);

        layoutSwipeRefresh = (SwipeRefreshLayout) findViewById(R.id.layout_swipe_refresh);
        gridMangaNewRelease = (RecyclerView) findViewById(R.id.grid_manga_new_release);
        progressLoading = (ProgressBar) findViewById(R.id.progress_loading);
        layoutSearch = (FrameLayout) findViewById(R.id.layout_search);
        inputSearch = (EditText) findViewById(R.id.input_search);
        animSlideUp = AnimationUtils.loadAnimation(this, R.anim.slide_up_anim);
        animSlideDown = AnimationUtils.loadAnimation(this, R.anim.slide_down_anim);
        newReleaseAdapter = new NewReleaseAdapter(this, this, storedDataService, mangaApiService);

        layoutSwipeRefresh.setColorSchemeColors(ContextCompat.getColor(this, R.color.colorAccent));
        gridMangaNewRelease.setAdapter(newReleaseAdapter);
        newReleaseAdapter.setOnListItemClickListener(this);
        inputSearch.setOnEditorActionListener(editorActionListener);
        inputSearch.addTextChangedListener(searchTextWatcher);
        layoutSwipeRefresh.setOnRefreshListener(pullRefreshListener);

        newReleaseAdapter.loadManga();
    }

    private TextView.OnEditorActionListener editorActionListener = new TextView.OnEditorActionListener() {
        @Override
        public boolean onEditorAction(TextView v, int actionId, KeyEvent event) {
            if (actionId == R.id.ime_action_search || actionId == EditorInfo.IME_ACTION_SEARCH) {
                newReleaseAdapter.filterList(inputSearch.getText().toString());
                hideKeyboard();
                return true;
            }
            return false;
        }
    };

    private TextWatcher searchTextWatcher = new TextWatcher() {
        @Override
        public void beforeTextChanged(CharSequence s, int start, int count, int after) {

        }

        @Override
        public void onTextChanged(CharSequence s, int start, int before, int count) {

        }

        @Override
        public void afterTextChanged(Editable editable) {
            newReleaseAdapter.filterList(editable.toString());
        }
    };

    private SwipeRefreshLayout.OnRefreshListener pullRefreshListener = new SwipeRefreshLayout.OnRefreshListener() {
        @Override
        public void onRefresh() {
            newReleaseAdapter.loadManga();
        }
    };

    @Override
    public void setTitle(CharSequence title) {
        super.setTitle(title);
        TextView textTitle = (TextView) findViewById(R.id.text_main_title);
        textTitle.setText(title);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        if (!isLoading) {
            MenuInflater inflater = getMenuInflater();
            inflater.inflate(R.menu.menu_main, menu);
        }
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public void onPrepare() {
        if (!layoutSwipeRefresh.isRefreshing())
            progressLoading.setVisibility(View.VISIBLE);
        isLoading = true;
        invalidateOptionsMenu();
    }

    @Override
    public void onSuccess() {
        layoutSwipeRefresh.setRefreshing(false);
        progressLoading.setVisibility(View.GONE);
        isLoading = false;
        invalidateOptionsMenu();
    }

    @Override
    public void onError(String errorMessage) {
        layoutSwipeRefresh.setRefreshing(false);
        progressLoading.setVisibility(View.GONE);
        showAlert("Error", errorMessage, "Reload", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialogInterface, int i) {
                newReleaseAdapter.loadManga();
            }
        });
    }

    @Override
    protected void onDestroy() {
        newReleaseAdapter.onParentDestroyed();
        super.onDestroy();
    }

    @Override
    public void onListItemClick(int position) {
        Intent intent = new Intent(this, ChapterListActivity.class);
        intent.putExtra(IntentKey.MANGA_DATA, newReleaseAdapter.getItem(position));
        startActivity(intent);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_search: {
                if (layoutSearch.getVisibility() == View.VISIBLE) {
                    item.setIcon(R.drawable.ic_search_white);
                    item.setTitle(R.string.search);
                    hideSearchLayout();
                } else {
                    item.setIcon(R.drawable.ic_close_white);
                    item.setTitle(R.string.close_search);
                    showSearchLayout();
                }
                return true;
            }
            case R.id.action_sort_by_release: {
                item.setChecked(true);
                newReleaseAdapter.setSortingOption(NewReleaseAdapter.SORT_BY_DATE);
                showToast("Sorted by release date.");
                return true;
            }
            case R.id.action_sort_by_title: {
                item.setChecked(true);
                newReleaseAdapter.setSortingOption(NewReleaseAdapter.SORT_BY_TITLE);
                showToast("Sorted by manga title.");
                return true;
            }
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    public void clearSearchClick(View view) {
        inputSearch.setText("");
    }

    private void hideSearchLayout() {
        animSlideUp.setAnimationListener(new Animation.AnimationListener() {
            @Override
            public void onAnimationStart(Animation animation) {

            }

            @Override
            public void onAnimationEnd(Animation animation) {
                inputSearch.clearFocus();
                layoutSearch.setVisibility(View.GONE);
                hideKeyboard();
            }

            @Override
            public void onAnimationRepeat(Animation animation) {

            }
        });
        layoutSearch.startAnimation(animSlideUp);
    }

    private void showSearchLayout() {
        layoutSearch.setVisibility(View.VISIBLE);
        animSlideDown.setAnimationListener(new Animation.AnimationListener() {
            @Override
            public void onAnimationStart(Animation animation) {

            }

            @Override
            public void onAnimationEnd(Animation animation) {
                inputSearch.requestFocus();
                showKeyboard(inputSearch);
            }

            @Override
            public void onAnimationRepeat(Animation animation) {

            }
        });
        layoutSearch.startAnimation(animSlideDown);
    }

    @Override
    public void onBackPressed() {
        if (layoutSearch.getVisibility() == View.VISIBLE)
            hideSearchLayout();
        else
            super.onBackPressed();
    }
}

File: app/src/main/java/com/bigscreen/mangindo/newrelease/NewReleaseActivity.kt
package com.bigscreen.mangindo.newrelease

import android.content.DialogInterface
import android.content.Intent
import android.os.Bundle
import androidx.core.content.ContextCompat
import android.text.Editable
import android.text.TextWatcher
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.animation.Animation
import android.view.animation.AnimationUtils
import android.view.inputmethod.EditorInfo
import android.widget.TextView
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.bigscreen.mangindo.R
import com.bigscreen.mangindo.base.BaseActivity
import com.bigscreen.mangindo.chapter.ChapterListActivity
import com.bigscreen.mangindo.common.IntentKey
import com.bigscreen.mangindo.common.extension.hideKeyboard
import com.bigscreen.mangindo.common.extension.showAlert
import com.bigscreen.mangindo.common.extension.showKeyboard
import com.bigscreen.mangindo.common.extension.showToast
import com.bigscreen.mangindo.common.listener.OnListItemClickListener
import com.bigscreen.mangindo.common.listener.OnLoadDataListener
import com.bigscreen.mangindo.repos.MangaRepository
import com.bigscreen.mangindo.stored.StoredDataService
import kotlinx.android.synthetic.main.activity_new_release.gridMangaNewRelease
import kotlinx.android.synthetic.main.activity_new_release.inputSearch
import kotlinx.android.synthetic.main.activity_new_release.layoutSearch
import kotlinx.android.synthetic.main.activity_new_release.layoutSwipeRefresh
import kotlinx.android.synthetic.main.activity_new_release.progressLoading
import javax.inject.Inject

class NewReleaseActivity : BaseActivity(), OnLoadDataListener, OnListItemClickListener {

    companion object {
        private const val IME_ACTION_SEARCH = 10
    }

    @Inject lateinit var storedDataService: StoredDataService
    @Inject lateinit var mangaRepository: MangaRepository

    private lateinit var newReleaseAdapter: NewReleaseAdapter
    private lateinit var animSlideUp: Animation
    private lateinit var animSlideDown: Animation

    private var isLoading = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        appDeps.inject(this)
        setContentView(R.layout.activity_new_release)
        supportActionBar?.setDisplayShowTitleEnabled(false)

        animSlideUp = AnimationUtils.loadAnimation(this, R.anim.slide_up_anim)
        animSlideDown = AnimationUtils.loadAnimation(this, R.anim.slide_down_anim)
        newReleaseAdapter = NewReleaseAdapter(this, this, storedDataService, mangaRepository)

        layoutSwipeRefresh.setColorSchemeColors(ContextCompat.getColor(this, R.color.colorAccent))
        gridMangaNewRelease.adapter = newReleaseAdapter
        newReleaseAdapter.listItemClickListener = this
        inputSearch.setOnEditorActionListener(editorActionListener)
        inputSearch.addTextChangedListener(searchTextWatcher)
        layoutSwipeRefresh.setOnRefreshListener(pullRefreshListener)

        newReleaseAdapter.loadManga()
    }

    private val editorActionListener = TextView.OnEditorActionListener { _, actionId, _ ->
        if (actionId == IME_ACTION_SEARCH || actionId == EditorInfo.IME_ACTION_SEARCH) {
            newReleaseAdapter.filterList(inputSearch.text.toString())
            hideKeyboard()
            return@OnEditorActionListener true
        }
        false
    }

    private val searchTextWatcher = object : TextWatcher {
        override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {
        }

        override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
        }

        override fun afterTextChanged(editable: Editable) {
            newReleaseAdapter.filterList(editable.toString())
        }
    }

    private val pullRefreshListener = SwipeRefreshLayout.OnRefreshListener { newReleaseAdapter.loadManga() }

    override fun setTitle(title: CharSequence) {
        super.setTitle(title)
        val textTitle = findViewById<TextView>(R.id.text_main_title)
        textTitle.text = title
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        if (!isLoading) {
            val inflater = menuInflater
            inflater.inflate(R.menu.menu_main, menu)
        }
        return super.onCreateOptionsMenu(menu)
    }

    override fun onPrepare() {
        if (!layoutSwipeRefresh.isRefreshing)
            progressLoading.visibility = View.VISIBLE
        isLoading = true
        invalidateOptionsMenu()
    }

    override fun onSuccess() {
        layoutSwipeRefresh.isRefreshing = false
        progressLoading.visibility = View.GONE
        isLoading = false
        invalidateOptionsMenu()
    }

    override fun onError(errorMessage: String) {
        layoutSwipeRefresh.isRefreshing = false
        progressLoading.visibility = View.GONE
        showAlert(
                "Error",
                errorMessage,
                "Reload",
                DialogInterface.OnClickListener {  _, _->
                    newReleaseAdapter.loadManga()
                }
        )
    }

    override fun onDestroy() {
        newReleaseAdapter.onParentDestroyed()
        super.onDestroy()
    }

    override fun onListItemClick(position: Int) {
        val intent = Intent(this, ChapterListActivity::class.java)
        intent.putExtra(IntentKey.MANGA_DATA, newReleaseAdapter.getItem(position))
        startActivity(intent)
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        when (item?.itemId) {
            R.id.action_search -> {
                if (layoutSearch.visibility == View.VISIBLE) {
                    item.setIcon(R.drawable.ic_search_white)
                    item.setTitle(R.string.search)
                    hideSearchLayout()
                } else {
                    item.setIcon(R.drawable.ic_close_white)
                    item.setTitle(R.string.close_search)
                    showSearchLayout()
                }
                return true
            }
            R.id.action_sort_by_release -> {
                item.isChecked = true
                newReleaseAdapter.setSortingOption(NewReleaseAdapter.SORT_BY_DATE)
                showToast("Sorted by release date.")
                return true
            }
            R.id.action_sort_by_title -> {
                item.isChecked = true
                newReleaseAdapter.setSortingOption(NewReleaseAdapter.SORT_BY_TITLE)
                showToast("Sorted by manga title.")
                return true
            }
            else -> return super.onOptionsItemSelected(item)
        }
    }

    fun clearSearchClick(@Suppress("UNUSED_PARAMETER") view: View) {
        inputSearch.setText("")
    }

    private fun hideSearchLayout() {
        animSlideUp.setAnimationListener(object : Animation.AnimationListener {
            override fun onAnimationStart(animation: Animation) {
            }

            override fun onAnimationEnd(animation: Animation) {
                inputSearch.clearFocus()
                layoutSearch.visibility = View.GONE
                this@NewReleaseActivity.hideKeyboard()
            }

            override fun onAnimationRepeat(animation: Animation) {
            }
        })
        layoutSearch.startAnimation(animSlideUp)
    }

    private fun showSearchLayout() {
        layoutSearch.visibility = View.VISIBLE
        animSlideDown.setAnimationListener(object : Animation.AnimationListener {
            override fun onAnimationStart(animation: Animation) {
            }

            override fun onAnimationEnd(animation: Animation) {
                inputSearch.requestFocus()
                this@NewReleaseActivity.showKeyboard(inputSearch)
            }

            override fun onAnimationRepeat(animation: Animation) {
            }
        })
        layoutSearch.startAnimation(animSlideDown)
    }

    override fun onBackPressed() {
        if (layoutSearch.visibility == View.VISIBLE)
            hideSearchLayout()
        else
            super.onBackPressed()
    }
}